anitya/lib/backends/github.py

```diff
@@ -132,36 +132,65 @@ def get_versions(cls, project):
                 )
             )
 
-        query = prepare_query(owner, repo, project.releases_only)
-
-        try:
-            headers = REQUEST_HEADERS.copy()
-            token = config["GITHUB_ACCESS_TOKEN"]
-            if token:
-                headers["Authorization"] = "bearer %s" % token
-            resp = http_session.post(
-                API_URL, json={"query": query}, headers=headers, timeout=60, verify=True
-            )
-        except Exception as err:
-            _log.debug("%s ERROR: %s" % (project.name, str(err)))
-            raise AnityaPluginException(
-                'Could not call : "%s" of "%s", with error: %s'
-                % (API_URL, project.name, str(err))
-            )
-
-        if resp.ok:
-            json = resp.json()
-        elif resp.status_code == 403:
-            _log.info("Github API ratelimit reached.")
-            raise RateLimitException(reset_time)
+        if project.latest_known_cursor:
+            # If we know about the cursor of the latest version, attempt to
+            # limit results to anything after it. Only if that fails, try
+            # without one.
+            cursor_attempts = (project.latest_known_cursor, None)
         else:
-            raise AnityaPluginException(
-                '%s: Server responded with status "%s": "%s"'
-                % (project.name, resp.status_code, resp.reason)
-            )
+            cursor_attempts = (None,)
+
+        for cursor in cursor_attempts:
```

Comment:
We are missing test, where the cursor_attempts is empty.

Reply:
`cursor_attempts` can't be empty, it is either `(project.latest_known_cursor, None)` or `(None,)`, per lines 139 & 141. It implements attempting retrieving versions after the latest known cursor at first, and if that fails (or no cursor is known), it retrieves the whole list.